include/ceed-hash.h

```diff
@@ -0,0 +1,149 @@
+  @param  h     Pointer to the hash table [khash_t(name)*]
+  @param  kvar  Variable to which key will be assigned
+  @param  vvar  Variable to which value will be assigned
+  @param  code  Block of code to execute
+ */
+#define kh_foreach(h, kvar, vvar, code) { khint_t __i;          \
+        for (__i = kh_begin(h); __i != kh_end(h); ++__i) {      \
+                if (!kh_exist(h,__i)) continue;                 \
+                (kvar) = kh_key(h,__i);                         \
+                (vvar) = kh_val(h,__i);                         \
+                code;                                           \
+        } }
+#endif /*kh_foreach*/
+
+#if !defined(kh_foreach_key)
+/*! @function
+  @abstract     Iterate over the keys in the hash table
+  @param  h     Pointer to the hash table [khash_t(name)*]
+  @param  kvar  Variable to which key will be assigned
+  @param  code  Block of code to execute
+ */
+#define kh_foreach_key(h, kvar, code) { khint_t __i;            \
+        for (__i = kh_begin(h); __i != kh_end(h); ++__i) {      \
+                if (!kh_exist(h,__i)) continue;                 \
+                (kvar) = kh_key(h,__i);                         \
+                code;                                           \
+        } }
+#endif /*kh_foreach_key*/
+
+#if !defined(kh_foreach_value)
+/*! @function
+  @abstract     Iterate over the values in the hash table
+  @param  h     Pointer to the hash table [khash_t(name)*]
+  @param  vvar  Variable to which value will be assigned
+  @param  code  Block of code to execute
+ */
+#define kh_foreach_value(h, vvar, code) { khint_t __i;          \
+        for (__i = kh_begin(h); __i != kh_end(h); ++__i) {      \
+                if (!kh_exist(h,__i)) continue;                 \
+                (vvar) = kh_val(h,__i);                         \
+                code;                                           \
+        } }
+#endif /*kh_foreach_value*/
+
+#define CeedHashGetValue(ht,k,v) ((v) = kh_value((ht),(k)))
+
+#define CeedHashMissing(ht,k) ((k) == kh_end((ht)))
+
+/* --- Thomas Wang integer hash functions --- */
+
+typedef khint32_t CeedHash32_t;
+typedef khint64_t CeedHash64_t;
+typedef khint_t   CeedHash_t;
+
+/* Thomas Wang's second version for 32bit integers */
+static inline CeedHash_t CeedHash_UInt32(CeedHash32_t key) {
+  key = ~key + (key << 15); /* key = (key << 15) - key - 1; */
+  key =  key ^ (key >> 12);
+  key =  key + (key <<  2);
+  key =  key ^ (key >>  4);
+  key =  key * 2057;        /* key = (key + (key << 3)) + (key << 11); */
+  key =  key ^ (key >> 16);
+  return key;
+}
+
+static inline CeedHash_t CeedHashInt(CeedInt key) {
+  return CeedHash_UInt32((CeedHash32_t)key);
+}
+
+static inline CeedHash_t CeedHashCombine(CeedHash_t seed, CeedHash_t hash) {
+  /* https://doi.org/10.1002/asi.10170 */
+  /* https://dl.acm.org/citation.cfm?id=759509 */
+  return seed ^ (hash + (seed << 6) + (seed >> 2));
+}
+
+#define CeedHashEqual(a,b) ((a) == (b))
```

Comment:
Never used?

Reply:
removed

Reply:
Ah, I had seen there were some unused definitions. I thought you put them for future reference. I wasn't sure indeed